[iOS] [Unified PDF] Make it possible to select words by long pressing

whsieh opened the pull request:
#### 6390ac13796720a15123cbb56cbe0971d9843e27
<pre>
[iOS] [Unified PDF] Make it possible to select words by long pressing
<a href="https://bugs.webkit.org/show_bug.cgi?id=284477">https://bugs.webkit.org/show_bug.cgi?id=284477</a>

Reviewed by Abrar Rahman Protyasha.

Work towards support for UIKit-based text interaction in PDFs when Unified PDF is enabled. This
patch only introduces the minimal support needed to allow long presses to select a single word in
a PDF document. See below for more details.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:

Remove some staging declarations that are no longer necessary, now that these enums are public API.

* Source/WebKit/Shared/EditorState.cpp:
(WebKit::operator&lt;&lt;):
* Source/WebKit/Shared/EditorState.h:

Move `isInPlugin` behind `PLATFORM(MAC)`. This flag is currently unused on iOS (no codepath sets it
to true).

* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(WebKit::WKSelectionDrawingInfo::WKSelectionDrawingInfo):
(WebKit::operator&lt;&lt;):

Remove dead code on iOS, now that `isInPlugin` is only available on macOS.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::requires):
(WebKit::PDFPluginBase::convertFromRootViewToPlugin const):
(WebKit::PDFPluginBase::convertFromPluginToRootView const):

Turn these into templated coordinate space conversion helper methods, and use a `CanMakeFloatRect`
concept here to call `mapRect` for both `IntRect` and `FloatRect` at compile time; otherwise, call
`mapPoint` for both `IntPoint` and `FloatPoint`.

(WebKit::PDFPluginBase::setSelectionRange):
(WebKit::PDFPluginBase::platformPopulateEditorStateIfNeeded const):

Move logic to set the `isInPlugin` flag into `PDFPluginBase`, from `WebPage`. This, as far as I
could tell, was only used on iOS prior to PDFKit integration, when the legacy PDF plugin was still
utilized. It&apos;s now only required for macOS. On iOS, the unified PDF plugin instead populates the
relevant members of `EditorState` and its associated structs.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::populateEditorStateIfNeeded const):
(WebKit::PDFPluginBase::convertFromRootViewToPlugin const): Deleted.
(WebKit::PDFPluginBase::convertFromPluginToRootView const): Deleted.

See above.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::continueTrackingSelection):

Drive-by fix: remove some now-unneeded explicit casts to `PDFSelectionGranularity`.

(WebKit::UnifiedPDFPlugin::setCurrentSelection):
(WebKit::UnifiedPDFPlugin::setSelectionRange):

Add plumbing to set the current selection at the root view point, with the given granularity. This
is invoked when long pressing to set the selection. We map the root view point into plugin
coordinates, hit-test the `PDFDocument` for a selection under the point, and then update the current
selection.

(WebKit::UnifiedPDFPlugin::platformPopulateEditorStateIfNeeded const):

Add logic to assemble an `EditorState` representing the selected text range in the PDF plugin, based
on `PDFSelection`. For now, this just handles the case where the selection is a single piece of
left-to-right text on a single line; in future patches, we&apos;ll need the ability to coalesce adjacent
rects, handle bidirectional text, and a number of other use cases.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::setSelectionRange):
(WebKit::PluginView::populateEditorStateIfNeeded const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):

Refactor this to defer to the PDF plugin if needed.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setFocusedFrameBeforeSelectingTextAtLocation):

Defer to the PDF plugin in the case where the focused or main frame is a PDF.

(WebKit::WebPage::setSelectionRange):

Canonical link: <a href="https://commits.webkit.org/287724@main">https://commits.webkit.org/287724@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc172a6733ba50dc1172495c2a828a09e83ee467

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80626 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34560 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85150 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31608 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68694 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7939 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62974 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20770 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83695 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53095 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73395 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43279 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50383 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30067 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28077 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86583 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7852 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71267 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69231 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70507 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17571 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14524 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13469 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7814 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7653 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11172 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9458 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -83,13 +83,20 @@ class WebFrame;
 
 enum class CheckValidRanges : bool { No, Yes };
 
+template<typename T>
+concept CanMakeFloatRect = requires(T t)
+{
+    { WebCore::FloatRect { t } } -> std::same_as<WebCore::FloatRect>;
+};
```
![image](https://github.com/user-attachments/assets/c38aa339-68a8-42a5-9477-a572abef5d67)

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -1192,6 +1173,22 @@
+
+    state.selectionIsNone = false;
+    state.selectionIsRange = true;
+    state.isInPlugin = true;
```
Hmm, this is sort of a misnomer now, I suppose. Maybe we should rename this to something more appropriate.

Right now this info percolates into `validRequestorForSendType:returnType:` (only validates send type against `legacyStringPasteboardType`) and for _something_ in WKContentViewInteraction, but can't tell...

whsieh replied:
> for something in WKContentViewInteraction

So I think that iOS-specific codepath can actually be removed altogether, since nothing sets the `isInPlugin` flag for iOS. On macOS, it's used for integration with system services (e.g. AppleScript) and Shortcuts, when obtaining selected content in the web view.

Perhaps we could just wrap this with a `PLATFORM(MAC)` compile-time flag and remove the dead code in `WKContentViewInteraction.mm`?

aprotyas replied:
Yeah I can do that if you send me a FIXME!

whsieh replied:
Ah, I just ended up cleaning that up as a part of this PR.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3874,6 +3877,95 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+        case TextGranularity::WordGranularity:
+            return [page selectionForWordAtPoint:pointInPage];
+        case TextGranularity::LineGranularity:
+            return [page selectionForLineAtPoint:pointInPage];
```
I think we can add CharacterGranularity support too, do you mind adding that or leaving a FIXME so we can evaluate if that even matters?

```suggestion
        case TextGranularity::CharacterGranularity:
            return [m_pdfDocument selectionFromPage:page.get() atPoint:pointInPage toPage:page.get() atPoint:pointInPage withGranularity:PDFSelectionGranularityCharacter];
        case TextGranularity::WordGranularity:
            return [page selectionForWordAtPoint:pointInPage];
        case TextGranularity::LineGranularity:
            return [page selectionForLineAtPoint:pointInPage];
```

aprotyas replied:
I wonder if `selectionFor[Word|Line]AtPoint:` behaves differently from `selectionFromPage:atPoint:toPage:atPoint:withGranularity:PDFSelectionGranularity[Word|Line]`

whsieh replied:
Oh, interesting — I can try using `-selectionFromPage:atPoint:toPage:atPoint:withGranularity:` here...

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3874,6 +3877,95 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+    if (!selection)
+        return;
+
+    setCurrentSelection(WTFMove(selection));
```
I think you want to unconditionally set the selection?

whsieh replied:
So I didn't want to clear the selection if it was an unsupported granularity, but perhaps I can unconditionally set it if I use the other method above.

aprotyas replied:
oh right, I see what you mean, but then we don't do the right thing if `selectionForWordAtPoint:` returns `nil` (when we have word granularity).

aprotyas replied:
Maybe return `std::optional<PDFSelection *>` in the lambda? `std::nullopt` for unsupported granularities, which we can use for an early return

whsieh replied:
Oh, I ended up falling back to a default granularity of `.character` (with a debug assertion), and then unconditionally call into `setCurrentSelection()` — I think that should make the logic a bit simpler to follow.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PluginView.h:
```diff
@@ -92,8 +93,11 @@ class PluginView final : public WebCore::PluginViewBase {
     void pluginScaleFactorDidChange();
 #if PLATFORM(IOS_FAMILY)
     void pluginDidInstallPDFDocument(double initialScaleFactor);
+    void setSelectionRange(WebCore::FloatPoint pointInRootView, WebCore::TextGranularity);
```
Can you add a forward decl for TextGranularity?

whsieh replied:
Will do!

whsieh committed:
[iOS] [Unified PDF] Make it possible to select words by long pressing
https://bugs.webkit.org/show_bug.cgi?id=284477

Reviewed by Abrar Rahman Protyasha.

Work towards support for UIKit-based text interaction in PDFs when Unified PDF is enabled. This
patch only introduces the minimal support needed to allow long presses to select a single word in
a PDF document. See below for more details.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:

Remove some staging declarations that are no longer necessary, now that these enums are public API.

* Source/WebKit/Shared/EditorState.cpp:
(WebKit::operator<<):
* Source/WebKit/Shared/EditorState.h:

Move `isInPlugin` behind `PLATFORM(MAC)`. This flag is currently unused on iOS (no codepath sets it
to true).

* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(WebKit::WKSelectionDrawingInfo::WKSelectionDrawingInfo):
(WebKit::operator<<):

Remove dead code on iOS, now that `isInPlugin` is only available on macOS.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::requires):
(WebKit::PDFPluginBase::convertFromRootViewToPlugin const):
(WebKit::PDFPluginBase::convertFromPluginToRootView const):

Turn these into templated coordinate space conversion helper methods, and use a `CanMakeFloatRect`
concept here to call `mapRect` for both `IntRect` and `FloatRect` at compile time; otherwise, call
`mapPoint` for both `IntPoint` and `FloatPoint`.

(WebKit::PDFPluginBase::setSelectionRange):
(WebKit::PDFPluginBase::platformPopulateEditorStateIfNeeded const):

Move logic to set the `isInPlugin` flag into `PDFPluginBase`, from `WebPage`. This, as far as I
could tell, was only used on iOS prior to PDFKit integration, when the legacy PDF plugin was still
utilized. It's now only required for macOS. On iOS, the unified PDF plugin instead populates the
relevant members of `EditorState` and its associated structs.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::populateEditorStateIfNeeded const):
(WebKit::PDFPluginBase::convertFromRootViewToPlugin const): Deleted.
(WebKit::PDFPluginBase::convertFromPluginToRootView const): Deleted.

See above.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::continueTrackingSelection):

Drive-by fix: remove some now-unneeded explicit casts to `PDFSelectionGranularity`.

(WebKit::UnifiedPDFPlugin::setCurrentSelection):
(WebKit::UnifiedPDFPlugin::setSelectionRange):

Add plumbing to set the current selection at the root view point, with the given granularity. This
is invoked when long pressing to set the selection. We map the root view point into plugin
coordinates, hit-test the `PDFDocument` for a selection under the point, and then update the current
selection.

(WebKit::UnifiedPDFPlugin::platformPopulateEditorStateIfNeeded const):

Add logic to assemble an `EditorState` representing the selected text range in the PDF plugin, based
on `PDFSelection`. For now, this just handles the case where the selection is a single piece of
left-to-right text on a single line; in future patches, we'll need the ability to coalesce adjacent
rects, handle bidirectional text, and a number of other use cases.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::setSelectionRange):
(WebKit::PluginView::populateEditorStateIfNeeded const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):

Refactor this to defer to the PDF plugin if needed.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setFocusedFrameBeforeSelectingTextAtLocation):

Defer to the PDF plugin in the case where the focused or main frame is a PDF.

(WebKit::WebPage::setSelectionRange):

Canonical link: https://commits.webkit.org/287724@main
